src/lib/components/Modal/__tests__/Modal.test.jsx

```diff
@@ -146,27 +159,125 @@ describe('functionality', () => {
     const spy = sinon.spy();
+    const ref = React.createRef();
     render((
       <Modal
-        {...mandatoryProps}
-        actions={[
-          {
-            disabled: true,
-            label: 'submit',
-            onClick: spy,
-            type: 'submit',
-          },
-        ]}
+        closeButtonRef={ref}
+        id="id"
       >
-        content
+        <ModalHead
+          title="title text"
+          closeButtonDisabled
+          closeButtonRef={ref}
+          onClose={spy}
+        />
       </Modal>
     ));
 
     action();
     expect(spy.called).toEqual(false);
   });
+
+  it('call primary action using `primaryButtonRef`', () => {
+    const spy = sinon.spy();
+    const ref = React.createRef();
+    render((
+      <Modal
+        primaryButtonRef={ref}
+        id="id"
+      >
+        <ModalFooter>
+          <Button
+            label="Submit"
+            onClick={spy}
+            ref={ref}
+            type="button"
+          />
+        </ModalFooter>
+      </Modal>
+    ));
+
+    userEvent.keyboard('{enter}');
+    expect(spy.calledOnce).toEqual(true);
+  });
+
+  it('do not call primary action using `primaryButtonRef ', () => {
+    const spy = sinon.spy();
+    const ref = React.createRef();
+    render((
+      <Modal
+        primaryButtonRef={ref}
+        id="id"
+      >
+        <ModalFooter>
+          <Button
+            disabled
+            label="Submit"
+            onClick={spy}
+            ref={ref}
+            type="button"
+          />
+        </ModalFooter>
+      </Modal>
+    ));
+
+    userEvent.keyboard('{enter}');
+    expect(spy.called).toEqual(false);
+  });
+
+  it.each([
+    [
+      <input />,
+      (rootElement) => {
+        expect(document.activeElement).toEqual(within(rootElement).getByRole('textbox'));
+      },
+    ],
+    [
+      <textarea />,
+      (rootElement) => {
+        expect(document.activeElement).toEqual(within(rootElement).getByRole('textbox'));
+      },
+    ],
+    [
+      (
+        <select>
+          <option>Option</option>
+        </select>
+      ),
+      (rootElement) => {
+        expect(document.activeElement).toEqual(within(rootElement).getByRole('combobox'));
+      },
+    ],
+  ])('autofocuses element (%#)', (child, assert) => {
+    const dom = render((
+      <Modal>
+        <ModalBody>
+          {child}
+        </ModalBody>
+      </Modal>
+    ));
+
+    assert(dom.container.firstChild);
+  });
+
+  it('autofocuses primary button if no input is present', () => {
+    const ref = React.createRef();
+    const { container } = render((
+      <Modal primaryButtonRef={ref}>
+        <ModalFooter>
+          <Button
+            label="Submit"
+            onClick={() => {}}
+            ref={ref}
+          />
+        </ModalFooter>
+      </Modal>
+    ));
+
+    expect(document.activeElement).toEqual(within(container.firstChild).getByRole('button'));
+  });
 });
```

Comment:
This one is a little bit complicated… For the record, I might have missed something here 👀. Other tests look good!

Reply:
You mean whole section go functional tests or just last test?

Reply:
I improved naming of last two tests of autofocus functionality if it helps.